### PR TITLE
[core] Pin object if it already exists

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

A worker can crash right after putting its return values into the object store. Then, the owner will receive the worker crashed error, but the return objects will still be in the remote object store. Later, if the task is retried, the worker will crash on [this line](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_actor_transport.cc#L105) because the object already exists.

Another way this can happen is if a task has multiple return values, and one of those return values is transferred to another node. If the task is later re-executed on that node, the task will fail because of the same error.

This PR fixes the crash so that:
1. If an object already exists, we try to pin that copy. Ideally, we should destroy the old copy and create the new one to make sure that metadata like the owner address is in sync, but this is pretty complicated to do right now.
2. If the pinning fails, we store an OBJECT_LOST error to throw to the application.
3. On the raylet, we check whether we already have the object pinned, and only subscribe to the owner's eviction message if the object is not pinned.
4. Also fixes bugs in the analogous case for `ray.put` (previously this would hang, now the application will receive an error if a `ray.put` object already exists).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Is this PR somehow mixed up with Pull timeout?

**Comment by rkooo567** [comment]
I've seen some changes from that PR reverted in this PR

**Comment by scv119** [comment]
test failure looks related?

**Comment by rkooo567** [comment]
```
test_task_output_inline_bytes_limit
```
failure seems related?

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_transport.cc:107]: Hmm does this mean if the local raylet is dead, the worker can just crash here? 

Can we handle this more gracefully without the error message?

**Review by rkooo567 — COMMENTED** [review]
Reviewed the first half. (except the local object manager path).

```
On the raylet, we check whether we already have the object pinned, and only subscribe to the owner's eviction message if the object is not pinned.
```

This one is needed because now we have non-owner who can pin the object right?

**Comment by rkooo567** [comment]
I think the risk of this change is these two assumptions no longer hold?

- There's only 1 pin object request to each object
- pin object request is only called from the owner

**Comment by stephanie-wang** [comment]
> I think the risk of this change is these two assumptions no longer hold?
> 
>     * There's only 1 pin object request to each object at max
> 
>     * pin object request is only called from the owner

Hmm actually I think the assumptions are pretty much the same as before. I don't think the second one was ever true? PinObjects should be called from the worker, not the owner.

The local object manager changes are for this case:
1. Worker creates object A, calls PinObjects on its local raylet.
2. Local raylet pins A, subscribes to object eviction message from owner.
3. Worker dies, so owner of A resubmits the task.
4. 2nd worker creates object A, calls PinObjects again on the local raylet. Local raylet errors because it already has a subscription for that object eviction.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc:2239]: I don't think it's really supposed to happen, but I think it's better to rely on higher-level failure handling (e.g., fetch timeout) than mark the object as lost.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:1843]: Sorry I don't understand this comment. Can you explain more?

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.cc:107]: I did this on purpose so that the caller can handle the error.

**Comment by rkooo567** [comment]
> Hmm actually I think the assumptions are pretty much the same as before. I don't think the second one was ever true? PinObjects should be called from the worker, not the owner.

Oh yeah sorry. I was confused between creator worker vs owner. So, even after this PR, there's only one worker that has in-flight pin request right? (so basically, there's no assumption change here).

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:1843]: Oh sorry. I didn't realize that `return_ptr ` is the `parameter[out]`. I just meant to explain how the function behaves when `  /// \param[out] return_object RayObject containing buffers to write results into.` is nullptr, thinking it was param[in]. You can ignore this message

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:2241]: Hmm can you comment that how the error is handled? Also, this means although the function itself returns "success", it is possible it could fail to pin it right? Can you also comment about this behavior in the docstring of this method? (e.g., Although the it returns success, the object may not be pinned if X happens or something like that)

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (rkooo567) [reply src/ray/core_worker/transport/direct_actor_transport.cc:107]: Oh I meant this will generate the ugly cpp stacktrace when this happens to the driver (agreed caller should handle the error). 

Maybe we can just do RAY_LOG(WARNING) and QuickExit() instead?

**Review by rkooo567 — APPROVED** [review]
LGTM If the assumption is not changed from before. Most of comments are related to docstring of new APIs. Also, should we create an issue to improve this protocol in the future (from pinning existing object to create a new object)?

File: src/ray/raylet/local_object_manager.cc
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:51]: Are we going to do this in a follow up? Also, is there any other case that owner address is needed from the raylet?

File: src/ray/raylet/local_object_manager.cc
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:262]: Can you comment what this block does?

**Comment by rkooo567** [comment]
> Hmm actually I think the assumptions are pretty much the same as before. I don't think the second one was ever true? PinObjects should be called from the worker, not the owner.

Ah yeah I was confused with SubscribeObjectEviction I think. So,

- PinObject basically is the called from the creator worker (same).
- And object has 1-in flight subscription to the owner to decide when to evict (same).
- But currently (which you will fix) it is possible the object subscribes to the wrong owner for eviction. If this happens, the object is lost unexpectedly now.

> A worker can crash right after putting its return values into the object store. Then, the owner will receive the worker crashed error, but the return objects will still be in the remote object store. Later, if the task is retried, the worker will crash on this line because the object already exists.

This seems like the owner address should be the same (since the owner re-exeuctes it).

> Another way this can happen is if a task has multiple return values, and one of those return values is transferred to another node. If the task is later re-executed on that node, the task will fail because of the same error.

And this is when we can have different owner address?

**Comment by stephanie-wang** [comment]
> > Hmm actually I think the assumptions are pretty much the same as before. I don't think the second one was ever true? PinObjects should be called from the worker, not the owner.
> 
> Ah yeah I was confused with SubscribeObjectEviction I think. So,
> 
>     * PinObject basically is the called from the creator worker (same).
> 
>     * And object has 1-in flight subscription to the owner to decide when to evict (same).
> 
>     * But currently (which you will fix) it is possible the object subscribes to the wrong owner for eviction. If this happens, the object is lost unexpectedly now.
> 
> 
> > A worker can crash right after putting its return values into the object store. Then, the owner will receive the worker crashed error, but the return objects will still be in the remote object store. Later, if the task is retried, the worker will crash on this line because the object already exists.
> 
> This seems like the owner address should be the same (since the owner re-exeuctes it).
> 
> > Another way this can happen is if a task has multiple return values, and one of those return values is transferred to another node. If the task is later re-executed on that node, the task will fail because of the same error.
> 
> And this is when we can have different owner address?

I think this is all right except for the last part. We can have a different owner address but same ObjectID because ObjectIDs are generated deterministically for tasks. I actually filed an issue here that explains it: #20713.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc:2241]: Will do. In general, the owner should be handling the failure, not the worker.

This behavior is the same as the usual return objects path, btw.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.cc:107]: Let's do this as part of a larger refactor. There are a lot of other RAY_CHECKs in this file and there's no way to access QuickExit() from here.

File: src/ray/raylet/local_object_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/local_object_manager.cc:51]: I think a better way might be to use nondeterministic ObjectIDs so we don't have the ID collision problem at all. Either way, it's out of scope for this PR.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (rkooo567) [reply src/ray/core_worker/transport/direct_actor_transport.cc:107]: Actually QuickExit is inside util.h. Is it still difficult in this case? I am okay with doing this in other PRs

**Comment by rkooo567** [comment]
So, according to the new policy, we need 2 approvers for 250+ lines of code PR. I will have a last look later today, but it was LGTM so far

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.cc:107]: Ah I see, my mistake! Thanks yeah I'll do it here.

**Pull request merged**
